components/script/dom/messageevent.rs

```diff
@@ -129,7 +151,11 @@ impl MessageEvent {
             false,
             message,
             DOMString::from(origin.unwrap_or("")),
-            source,
+            source
+                .map(|source| {
+                    WindowProxyOrMessagePortOrServiceWorker::WindowProxy(DomRoot::from_ref(source))
```

Comment:
I think you might want to instead update the `source` argument to be a `WindowProxyOrMessagePortOrServiceWorker`, and then set it at the call site, at https://github.com/servo/servo/blob/c89676be1d4b0e727b793f39cfb2d5db006ff1a8/components/script/dom/globalscope.rs#L534 and https://github.com/servo/servo/blob/c89676be1d4b0e727b793f39cfb2d5db006ff1a8/components/script/dom/dedicatedworkerglobalscope.rs#L479

Then you could also close https://github.com/servo/servo/issues/24454

Reply:
[The spec](https://html.spec.whatwg.org/multipage/web-messaging.html#message-port-post-message-steps) does not set the `source` attribute, should the spec be updated?

Reply:
That's a good point. I double-checked Gecko's implementation and it also appears to have a null source for message events triggered by MessagePort.postMessage, which I did not expect.

Reply:
Maybe it's because for two MessagePorts in separate event loops there is no reasonable object to provide for the source argument.

Reply:
https://html.spec.whatwg.org/multipage/comms.html#dom-messageevent-source suggests that the MessagePort is only used for shared worker connect events, which makes more sense to me.

Reply:
You're right, it seems to only be used for shared-workers in step 27 of https://html.spec.whatwg.org/multipage/workers.html#worker-processing-model